internal/services/devcenter/dev_center_project_environment_type.go

```diff
@@ -0,0 +1,406 @@
+			if err != nil {
+				return err
+			}
+
+			var model DevCenterProjectEnvironmentTypeResourceModel
+			if err := metadata.Decode(&model); err != nil {
+				return fmt.Errorf("decoding: %+v", err)
+			}
+
+			resp, err := client.ProjectEnvironmentTypesGet(ctx, *id)
+			if err != nil {
+				return fmt.Errorf("retrieving %s: %+v", id, err)
+			}
+
+			properties := resp.Model
+			if properties == nil {
+				return fmt.Errorf("retrieving %s: model was nil", id)
+			}
+
+			if metadata.ResourceData.HasChange("creator_role_assignment_roles") {
+				properties.Properties.CreatorRoleAssignment.Roles = expandDevCenterProjectEnvironmentTypeCreatorRoleAssignmentRoles(model.CreatorRoleAssignmentRoles)
+			}
+
+			if metadata.ResourceData.HasChange("deployment_target_id") {
+				properties.Properties.DeploymentTargetId = pointer.To(model.DeploymentTargetId)
+			}
+
+			if metadata.ResourceData.HasChange("identity") {
+				identity, err := identity.ExpandSystemAndUserAssignedMapFromModel(model.Identity)
+				if err != nil {
+					return err
+				}
+				properties.Identity = identity
+			}
+
+			if metadata.ResourceData.HasChange("user_role_assignment") {
+				properties.Properties.UserRoleAssignments = expandDevCenterProjectEnvironmentTypeUserRoleAssignment(model.UserRoleAssignment)
+			}
+
+			if metadata.ResourceData.HasChange("tags") {
+				properties.Tags = pointer.To(model.Tags)
+			}
+
+			if _, err := client.ProjectEnvironmentTypesCreateOrUpdate(ctx, *id, *properties); err != nil {
+				return fmt.Errorf("updating %s: %+v", *id, err)
+			}
+
+			return nil
+		},
+	}
+}
+
+func (r DevCenterProjectEnvironmentTypeResource) Delete() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.DevCenter.V20230401.EnvironmentTypes
+
+			id, err := environmenttypes.ParseEnvironmentTypeID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			if _, err := client.ProjectEnvironmentTypesDelete(ctx, *id); err != nil {
+				return fmt.Errorf("deleting %s: %+v", *id, err)
+			}
+
+			return nil
+		},
+	}
+}
+
+func expandDevCenterProjectEnvironmentTypeCreatorRoleAssignmentRoles(input []string) *map[string]environmenttypes.EnvironmentRole {
+	if len(input) == 0 {
+		return nil
+	}
+
+	result := map[string]environmenttypes.EnvironmentRole{}
+
+	for _, v := range input {
+		if _, exists := result[v]; exists {
+			fmt.Println("creator role assignment role is duplicate in the list")
+		}
+
+		result[v] = environmenttypes.EnvironmentRole{}
+	}
+
+	return &result
+}
+
+func expandDevCenterProjectEnvironmentTypeUserRoleAssignment(input []DevCenterProjectEnvironmentTypeUserRoleAssignment) *map[string]environmenttypes.UserRoleAssignment {
+	if len(input) == 0 {
+		return nil
+	}
+
+	result := map[string]environmenttypes.UserRoleAssignment{}
+
+	for _, v := range input {
+		if _, exists := result[v.UserId]; exists {
```

Comment:
shall we validate the duplication in `customizeDiff`?

Reply:
When there is a sub property that is defined as TypeSet, customizeDiff cannot work normally with it. For example, TF cannot get the value of `user_id` in customizeDiff.